### DISCUSSION TITLE
How to create a status bar?

### DISCUSSION THREAD

**Discussion by 0x7FFFFFFFFFFFFFFF** [gyscos/cursive#677] [Q&A] [answered]
How can I create a status bar like this one? It's always at the bottom and always top most. 

![image](https://user-images.githubusercontent.com/39968/192096777-262446d4-6ec7-421a-ba12-625229def6d1.png)

Can I use cursive to achieve this? Or must I have to use a lower level lib like crossterm? Thanks.

**Comment by gyscos** [comment, marked as answer]
Hi, and thanks for the question!

To create a status bar, you can add a transparent full-screen layer at the back of your StackView, which only draws something at the bottom.
For example, you can use a `FixedLayout` to manually specify the location of a `TextView`, and use `OnLayoutView` to update this location when the terminal is resized:

```rust
use cursive::traits::{Nameable, Resizable};
use cursive::view::View;
use cursive::views::{FixedLayout, Layer, OnLayoutView, TextView};
use cursive::{Rect, Vec2};

fn main() {
    let mut siv = cursive::default();

    siv.screen_mut().add_transparent_layer(
        OnLayoutView::new(
            FixedLayout::new().child(
                Rect::from_point(Vec2::zero()),
                Layer::new(TextView::new("Status").with_name("status"))
                    .full_width(),
            ),
            |layout, size| {
                layout.set_child_position(
                    0,
                    Rect::from_size((0, size.y - 1), (size.x, 1)),
                );
                layout.layout(size);
            },
        )
        .full_screen(),
    );

    siv.run();
}
```

(Using a `FixedLayout` is just one way to achieve this; a custom view or a `Canvas` may also work here.)

**Reply by stIncMale** [reply]
Thank you for the answer, @gyscos 

> add a transparent full-screen layer at the back of your StackView, which only draws something at the bottom

If such a transparent layer is not at the top of the stack, then it is drawn over by other layers, so it has to be at the top. However, if it's at the top, then it prevents mouse/keyboard events from propagating to the layers down the stack. Is there an easy way for such a layer to let events propagate?

**Reply by gyscos** [reply]
We don't currently support having a "front/top" layer without intercepting input (it's planned, but not in the very short-term).

Because it is at the bottom of the screen, it is out of the way of regular centered views. You would need a view to be full-screen to hide it.

To prevent that, you have some options:
* Never use fullscreen views in front of the status bar. When you would want a fullscreen view, instead use the same trick as for the StatusBar, to have a view that is _almost_ fullscreen but with a transparent row at the bottom, leaving the status bar visible.
* Have a single layer with a vertical layout: the status bar at the bottom, and a `StackView` above. Then you would just use _that_ `StackView` rather than the default one at the root.
  * This is the cleanest solution, and we may want to make it easier to use, ideally without having to replace the root StackView.

**Reply by stIncMale** [reply]
Thank you! I tried your second (cleanest) solution. It works, but the result gets cluttered with shadows. Disabling shadows via `Cursive.update_theme` makes things better, but does not remove borders preserved for shadows, which is not good. Besides, I want shadows on some views, but not on the status bar. I tried changing which views have shadows and which don't via `ThemedView`/`ShadowView` - these don't work as I'd expect them to work (in most cases they have no visible effect).
